compiler/rustc_codegen_ssa/src/debuginfo/type_names.rs

```diff
@@ -179,27 +188,27 @@ pub fn push_debuginfo_type_name<'tcx>(
                 push_item_name(tcx, principal.def_id, qualified, output);
                 push_generic_params_internal(tcx, principal.substs, false, output, visited);
             } else {
-                for single_trait in trait_data.iter() {
-                    match tcx.normalize_erasing_late_bound_regions(
-                        ty::ParamEnv::reveal_all(),
-                        single_trait,
-                    ) {
-                        ty::ExistentialPredicate::Trait(trait_ref) => {
-                            push_item_name(tcx, trait_ref.def_id, true, output);
-                            push_generic_params_internal(
-                                tcx,
-                                trait_ref.substs,
-                                false,
-                                output,
-                                visited,
-                            );
-                        }
+                let mut traits: Vec<_> = trait_data
+                    .iter()
+                    .map(|single_trait| {
+                        tcx.normalize_erasing_late_bound_regions(
+                            ty::ParamEnv::reveal_all(),
+                            single_trait,
+                        )
+                    })
+                    .collect();
+                traits.sort();
```

Comment:
I noticed that I kept getting tests failures for the `let no_principal_trait = (box 0_isize) as Box<dyn Send + Sync>;` scenario where the `Send` and `Sync` trait were switching their order in the output, so I adding sorting to ensure a stable order.

Reply:
Hm, OK. We should try to find out why the order is not deterministic. This might indicate a bug further up the pipeline.

If it turns out that we really need to sort the items here, we must sort them in a stable order. Relying on the compiler generated `Ord` implementation would mean that item order is defined by the numeric values of compiler internal IDs (like the `DefId` in  `ExistentialPredicate::AutoTrait(def_id)`).

Reply:
It looks like this list should already be in a stable order:
https://github.com/rust-lang/rust/blob/2312ff1a850db449b79fd3c4b215395cd2598c25/compiler/rustc_typeck/src/astconv/mod.rs#L1405

So there's something fishy here.

Reply:
Ok, I'll remove the sorting - the order changing only seemed to happen when I rebased, so it might be possible that something changed and then changed back with each rebase?

Reply:
Looks like the pretty printer had to use the same workaround:
https://github.com/rust-lang/rust/blob/e4a60327063e82413eed50a10df3b7d19b77bda0/compiler/rustc_middle/src/ty/print/pretty.rs#L863

If I understand correctly, `DefPathHash` is supposed to be stable if the environment is stable (e.g., between incremental compilations) but doesn't make any guarantees otherwise (e.g., changes to compiler version, targets)